### ISSUE TITLE
Pre-release standardization: tracking issue

### ISSUE THREAD

**Issue by yannham** [tweag/nickel#462]
This issue is meant to track a bunch of important practical usability aspects, that have been mostly discussed in issues, meetings or design documents, sometimes extensively, but without reaching a clear conclusion. The goal is to go over them and decide once and for all for the first release (or, at least, to consciously and willfully declare them out of scope for the first release, and to be debated later).

## Issues

- [x] Should we drop the `fun` keyword for something else, like just the arrow `foo bar => functionBody`, with parentheses `(foo bar) => body` or use a character as in Haskell `\foo bar => body`? Also, should we have an ML-like function syntax avoiding the anonymous function syntax most of the time: `let f x y = x + y in` ?
      Related: #207, in particular https://github.com/tweag/nickel/issues/207#issuecomment-900271443)
- [x] Should we allow a record to reference fields possibly coming from "sibling" from future merges once #330 lands, as in `{foo = 1} & {bar = foo}`? It introduces a form of dynamic binding. Alternatives were to force the second record to specify it expects a `foo` field to be defined in some way or another, as `{ bar = foo, foo | Dyn}`.
      Related: #83, #330, #459
- [x] Should we, and if yes how, differentiate custom contracts from types, both at definition and at usage?
      Related: #295, #318
- [x] Should we have a type hole syntax, to let the typechecker fill the gap, as in `let foo : _ = ... in ...` ?
      Related: #226, #430
- [ ] Should we change the file extension?
      Related: #357
- [x] Should we have a block let-binding syntax?
     Related: #218
- [ ] Standard library stabilization
     Related: #321 

**ADDENDUM**

- [x] Should we have top-level let-bindings?
      Related: #95
- [x] Case convention for values/types/contracts/stdlib

**Comment by yannham** [comment]
The first release has passed, and most of those points have been decided. Open issues remain for the others.
